modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/components/A11y/A11yChecker.ts

```diff
@@ -0,0 +1,398 @@
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import axe, {AxeResults, RunOptions} from 'axe-core';
+
+declare global {
+	interface Window {
+		requestIdleCallback(callback: Function): number;
+		cancelIdleCallback(handle: number): void;
+	}
+}
+
+type Task<T> = {
+	id: number;
+	callback: Function;
+	target: T;
+};
+
+type Selector<T> = (
+	this: void,
+	value: T,
+	index: number,
+	obj: Array<T>
+) => unknown;
+
+class Queue<T> {
+	queue: Array<T> = [];
+
+	enqueue(value: T) {
+		return this.queue.push(value);
+	}
+
+	dequeue() {
+		return this.queue.shift();
+	}
+
+	peek() {
+		return this.queue.length === 0 ? undefined : this.queue[0];
+	}
+
+	has(selector: Selector<T>) {
+		return this.queue.find(selector);
+	}
+
+	size() {
+		return this.queue.length;
+	}
+}
+
+/**
+ * Scheduler deals with scheduling a callback to A11y that runs axe-core at the
+ * end, many of the logics and algorithms in the scheduler are restricted to
+ * A11y but partially generic.
+ *
+ * The implementation is based on Queue. Each new scheduled callback is added
+ * end of the queue, the scheduler works with the main callback which is
+ * responsible for emptying the queue and executing the queue callbacks in the
+ * work loop respecting the frame deadline.
+ *
+ * Main thread
+ * ------------------------------------------------------------------------->
+ *
+ * +------------+ ++ ++ ++ +--------+            +--------+    +------+
+ * |    Busy    | || || || |  Idle  |            |  Busy  |    | Idle |
+ * +------------+ ++ ++ ++ +--------+            +--------+    +------+
+ *
+ *                         +----------+ +---+                  +-----------+
+ *                         | Job Exec.| |   |   Yield to host  | Job Exec. |
+ *                         +----------+ +---+ <--------------> +-----------+
+ *                        |                  |
+ *                        +------------------+
+ *                              Deadline
+ *
+ * When the main thread is idle the main callback is called to empty the queue
+ * then a work loop is executed to ensure that we can execute more than one
+ * callback until the defined deadline but it is interrupted to avoid loss
+ * of frame and at the end of each execution it defines the next frame phase
+ * that avoids abusing cycles and yields up control of the main thread to
+ * the host. The main callback will be rescheduled again until the queue is
+ * empty.
+ */
+class Scheduler<T> {
+	private taskIdCounter: number = 0;
+	private readonly queue: Queue<Task<T>>;
+	private currentTask: Task<T> | undefined = undefined;
+	private scheduledHostHandle: number = 0;
+	private isCallbackScheduled: boolean = false;
+	private isPerformingWork: boolean = false;
```

Comment:
I think TS can infer the types of these primitives from their default value, so there is no need to declare the types.

```ts
class Scheduler {
	private taskIdCounter = 0;
	private isCallbackScheduled = false;

    test() {
        this.taskIdCounter; // type is a number
        this.isCallbackScheduled; // type is a boolean
    }
}
```